table/tables/cache.go

```diff
@@ -30,74 +34,148 @@ var _ table.CachedTable = &cachedTable{}
+	const defaultLeaseDuration time.Duration = 3 * time.Second
+	physicalTime := oracle.GetTimeFromTS(ts)
+	lease := oracle.GoTimeToTS(physicalTime.Add(defaultLeaseDuration))
+	return lease
 }
 
 func (c *cachedTable) TryGetMemcache(ts uint64) (kv.MemBuffer, bool) {
-	c.mu.RLock()
-	defer c.mu.RUnlock()
-	if c.isReadFromCache(ts) {
-		return c.MemBuffer, true
+	tmp := c.cacheData.Load()
+	if tmp == nil {
+		return nil, false
+	}
+	data := tmp.(*table.CacheData)
+	if data.Lease > ts {
+		return data.MemBuffer, true
 	}
 	return nil, false
 }
-func (c *cachedTable) isReadFromCache(ts uint64) bool {
-	// If first read cache table. directly return false, the backend goroutine will help us update the lock information
-	// and read the data from the original table at the same time
-	// TODO : Use lease and ts judge whether it is readable.
-	// TODO : If the cache is not readable. MemBuffer become invalid.
-	return c.MemBuffer != nil
-}
+
+var mockStateRemote = struct {
+	Ch   chan remoteTask
+	Data *mockStateRemoteData
+}{}
 
 // NewCachedTable creates a new CachedTable Instance
 func NewCachedTable(tbl *TableCommon) (table.Table, error) {
-	return &cachedTable{
+	if mockStateRemote.Data == nil {
+		mockStateRemote.Data = newMockStateRemoteData()
+		mockStateRemote.Ch = make(chan remoteTask, 100)
+		go mockRemoteService(mockStateRemote.Data, mockStateRemote.Ch)
+	}
+	ret := &cachedTable{
 		TableCommon: *tbl,
-	}, nil
+		handle:      &mockStateRemoteHandle{mockStateRemote.Ch},
+	}
+
+	return ret, nil
 }
 
-func (c *cachedTable) loadDataFromOriginalTable(ctx sessionctx.Context) error {
+func (c *cachedTable) loadDataFromOriginalTable(ctx sessionctx.Context, lease uint64) (kv.MemBuffer, error) {
 	prefix := tablecodec.GenTablePrefix(c.tableID)
 	txn, err := ctx.Txn(true)
 	if err != nil {
-		return err
+		return nil, err
 	}
+	if txn.StartTS() >= lease {
+		return nil, errors.New("the loaded data is outdate for caching")
+	}
+
 	buffTxn, err := ctx.GetStore().BeginWithOption(tikv.DefaultStartTSOption().SetStartTS(0))
 	if err != nil {
-		return err
+		return nil, err
 	}
+
 	buffer := buffTxn.GetMemBuffer()
 	it, err := txn.Iter(prefix, prefix.PrefixNext())
 	if err != nil {
-		return err
+		return nil, err
 	}
 	defer it.Close()
-	if !it.Valid() {
-		return nil
-	}
 	for it.Valid() && it.Key().HasPrefix(prefix) {
 		value := it.Value()
 		err = buffer.Set(it.Key(), value)
 		if err != nil {
-			return err
+			return nil, err
 		}
 		err = it.Next()
 		if err != nil {
-			return err
+			return nil, err
 		}
 	}
-	c.mu.Lock()
-	c.MemBuffer = buffer
-	c.mu.Unlock()
-	return nil
+
+	return buffer, nil
 }
 
 func (c *cachedTable) UpdateLockForRead(ctx sessionctx.Context, ts uint64) error {
-	// Now only the data is re-load here, and the lock information is not updated. any read-lock information update will in the next pr.
-	err := c.loadDataFromOriginalTable(ctx)
+	// Load data from original table and the update lock information.
```

Comment:
Change `UpdateLockForRead(ctx sessionctx.Context, ts uint64)` to `UpdateLockForRead(store kv.Storage, ts uint64)` to avoid misuse for `sessionctx.Context` that can cause data race. You can do it in next pr.

Reply:
okay. i will modify it in the next pr.